hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java

```diff
@@ -720,29 +759,38 @@ private int checkMissingStates() {
     }
 
     /*
-        Invoked when a data is added on a different node and this data should be propagated
+        Data incoming from a different node.  This has already been globally filtered but not locally filtered.
+        It does not need to be re-propagated.
+
+        We allow concurrent threads to make this call in order to process distributed data in parallel. We
+        use synchronized blocks to protect pendingData.
      */
     @Override
+    @Lock(LockType.READ)
     public void onNewData(Data data) {
-        addPendingData(data);
+        sendData(data);
```

Comment:
Here it is the problem.
`onNewData` receives the notification of the distribution of the data.

But if we invoke to
```java
    public void sendData(TreeSet<Data> data) {
        if (data == null) {
            throw new IllegalArgumentException("Data must be not null");
        }
        if (data.isEmpty()) {
            return;
        }

        addData(data);

        if (distributed) {
            partitionManager.notifyData(new ArrayList<>(data));
        }
    }
```
Then we are looping again the data into the cluster, which is not good.

So, `onNewData` should just add the received data into the pendingData for this node.

The cluster error was generated due duplicated data in a node coming from others nodes.